wagtail/models/sites.py

```diff
@@ -215,55 +286,79 @@ def get_site_root_paths():
-                                root_page.locale.language_code,
-                            )
-                            for root_page in site.root_page.get_translations(
-                                inclusive=True
-                            ).select_related("locale")
-                        ]
-                    )
-                else:
-                    result.append(
+        NOTE: Unless the `WAGTAIL_PER_THREAD_SITE_CACHING` setting has been set to `False`, the
+        return value will be cached for the current thread.
+        """
+        caching_enabled = per_thread_site_caching_enabled()
+        if caching_enabled:
+            cached_result = cls._get_cached_site_root_paths()
+            if cached_result is not None:
+                return cached_result
+
+        result = []
+        for site in Site.objects.get_all():
+            if getattr(settings, "WAGTAIL_I18N_ENABLED", False):
+                result.extend(
+                    [
                         SiteRootPath(
                             site.id,
-                            site.root_page.url_path,
+                            root_page.url_path,
                             site.root_url,
-                            site.root_page.locale.language_code,
+                            root_page.locale.language_code,
                         )
+                        for root_page in site.root_page.get_translations(
+                            inclusive=True
+                        ).select_related("locale")
+                    ]
+                )
+            else:
+                result.append(
+                    SiteRootPath(
+                        site.id,
+                        site.root_page.url_path,
+                        site.root_url,
+                        site.root_page.locale.language_code,
                     )
+                )
 
-            cache.set(
-                SITE_ROOT_PATHS_CACHE_KEY,
-                result,
-                3600,
-                version=SITE_ROOT_PATHS_CACHE_VERSION,
-            )
+        # Convert to a tuple for weak reference support
+        result = tuple(result)
 
-        else:
-            # Convert the cache result to a list of SiteRootPath tuples, as some
-            # cache backends (e.g. Redis) don't support named tuples.
-            result = [SiteRootPath(*result) for result in result]
+        if caching_enabled:
+            # A copy is cached to prevent mutation and creation of
+            # strong references to cached values
+            _site_root_paths_cache.value = deepcopy(result)
 
         return result
 
     @staticmethod
-    def clear_site_root_paths_cache():
-        cache.delete(SITE_ROOT_PATHS_CACHE_KEY, version=SITE_ROOT_PATHS_CACHE_VERSION)
+    def _get_cached_site_root_paths() -> Union[List[SiteRootPath], None]:
+        result = getattr(_site_root_paths_cache, "value", None)
+        if result is not None:
```

Comment:
Nit: viva la walrus!

```suggestion
        if result := getattr(_site_root_paths_cache, "value", None):
```

Reply:
Beware suggesting use of walrus operators where the difference between `None` and other falsey values are important :)

Reply:
They are not mutually exclusive, e.g.

```py
if result := getattr(_site_root_paths_cache, "value", None) is not None:
```

Reply:
@laymonage I'm not sure if this is on purpose or not, but I think that example would result in `result` ending up as a `bool` value, when what we really want is the `list` (empty or not) from the cache, if there is one.

Reply:
Oops yep, you're right! It would need a parenthesis to work:

```py
if (result := getattr(_site_root_paths_cache, "value", None)) is not None:
```

at which point I think you're better off without the walrus 😄